include/deal.II/lac/solver_idr.h

```diff
@@ -0,0 +1,495 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2000 - 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_solver_idr_h
+#define dealii_solver_idr_h
+
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/logstream.h>
+#include <deal.II/base/signaling_nan.h>
+#include <deal.II/base/subscriptor.h>
+#include <deal.II/base/utilities.h>
+
+#include <deal.II/lac/lapack_full_matrix.h>
+#include <deal.II/lac/solver.h>
+#include <deal.II/lac/solver_control.h>
+
+#include <boost/random.hpp>
+
+#include <cmath>
+
+DEAL_II_NAMESPACE_OPEN
+
+/*!@addtogroup Solvers */
+/*@{*/
+
+namespace internal
+{
+  /**
+   * A namespace for a helper class to the IDR(s) solver.
+   */
+  namespace SolverIDRImplementation
+  {
+    /**
+     * Class to hold temporary vectors whose size depends on
+     * the solver parameter s.
+     */
+    template <typename VectorType>
+    class TmpVectors
+    {
+    public:
+      /**
+       * Constructor. Prepares an array of @p VectorType of length @p s_param.
+       */
+      TmpVectors(const unsigned int s_param, VectorMemory<VectorType> &vmem);
+
+      /**
+       * Destructor. Delete all allocated vectors.
+       */
+      ~TmpVectors() = default;
+
+      /**
+       * Get vector number @p i. If this vector was unused before, an error
+       * occurs.
+       */
+      VectorType &operator[](const unsigned int i) const;
+
+      /**
+       * Get vector number @p i. Allocate it if necessary.
+       *
+       * If a vector must be allocated, @p temp is used to reinit it to the
+       * proper dimensions.
+       */
+      VectorType &
+      operator()(const unsigned int i, const VectorType &temp);
+
+      /**
+       * Return size of data vector.
```

Comment:
is this the number of vectors stored?

Reply:
It is the number of vectors stored in the created object. I use 3 of these objects in the solve() function: U, G, and Q, each with `s` vectors. This class (like the temporary vectors in GMRES, I copied this class from solver_gmres.h) allows me to organize the vectors which depend on `s`.

Reply:
This is actually from the GMRES implementation and not needed here. I am removing it.